### PR TITLE
Add status property to units and applications

### PR DESCRIPTION
**Pull request by dshcherb**
In order to allow status assignment for units and applications a status
descriptor is used.

Status for remote units and applications is set to unknown. While it is
possible to retrieve remote unit status via goal-state (but not
remote application status), this change does not aim to address this
case.

### CODE REVIEW

File: juju/model.py [hunk 1]
```diff
@@ -43,8 +45,11 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
+
     def get_unit(self, unit_name):
-        return self._cache.get(Unit, unit_name)
+        return self._cache.get(Unit, unit_name, self._is_local_unit(unit_name), self._backend)
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/model.py]: Thanks, I changed it to address this.

File: test/test_model.py [hunk 2]
```diff
@@ -61,12 +62,25 @@ def config_get(self):
...
+        self.status_set_calls.append((app, status, message))
+
 
 class TestModel(unittest.TestCase):
     def setUp(self):
         self.backend = TestModelBackend()
         self.model = juju.model.Model('myapp/0', ['db0', 'db1', 'db2'], self.backend)
 
+        juju.model.Unit.status = juju.model.EntityStatus()
+        juju.model.Application.status = juju.model.EntityStatus()
```
Comment (johnsca) [inline test/test_model.py]: I take it these are to clear the cached values between tests?  Probably worth adding a comment here.

File: juju/model.py [hunk 3]
```diff
@@ -43,32 +45,88 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
```
Comment (niemeyer) [inline juju/model.py]: It's not worth having a function that looks like
```
        unit._is_local_unit(some_unit.name)
```
When its single job is replacing the line
```
        unit._local_unit_name == some_unit.name
```

File: juju/model.py [hunk 4]
```diff
@@ -3,12 +3,14 @@
...
 from collections.abc import Mapping, MutableMapping
 from subprocess import run, PIPE, CalledProcessError
+from enum import Enum, unique
 
 
 class Model:
     def __init__(self, local_unit_name, relation_names, backend):
         self._cache = ModelCache()
         self._backend = backend
+        self._local_unit_name = local_unit_name
```
Comment (niemeyer) [inline juju/model.py]: It's not worth having an attribute that looks like
```
        self._local_unit_name
```
when its single job is replacing the line
```
        self.unit.name
```

File: juju/model.py [hunk 5]
```diff
@@ -43,32 +45,88 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
+
     def get_unit(self, unit_name):
-        return self._cache.get(Unit, unit_name)
+        return self._cache.get(CacheKey(Unit, unit_name), unit_name, self._is_local_unit(unit_name), self._backend, self._cache)
```
Comment (niemeyer) [inline juju/model.py]: I'm a bit confused by these changes, as this seems unrelated to the specific purpose of this PR. Is this merging other PRs, and if so can we get it clean with only the specific modifications for status?

File: juju/model.py [hunk 6]
```diff
@@ -43,32 +45,88 @@ def get_relation(self, relation_name, relation_id=None):
...
         return entity
 
+class EntityStatus:
+    """A helper class for getting and setting entity (application or unit) status."""
+
+    def __init__(self):
+        self._cached = None
+
+    def __get__(self, instance, owner):
+        is_app = isinstance(instance, Application)
```
Comment (niemeyer) [inline juju/model.py]: It doesn't make much sense to have a mixin class that actually knows about the exact details of all classes where it's used. It all becomes a single big and complex class that pretends to be a set of composible types.

File: juju/model.py [hunk 7]
```diff
@@ -231,6 +289,54 @@ class RelationNotFound(ModelError):
...
+    def __init__(self, message):
+        super().__init__('waiting', message)
+
+@unique
+class StatusTypes(Enum):
+    active = Active
+    blocked = Blocked
+    maintenance = Maintenance
+    unknown = Unknown
+    waiting = Waiting
```
Comment (niemeyer) [inline juju/model.py]: That seems bogus. This is an example of Enum usage, from the PEP:
```
>>> from enum import Enum
>>> class Color(Enum):
...     RED = 1
...     GREEN = 2
...     BLUE = 3
```
What you have there are attributes that are actually types.

Comment (johnsca) [reply juju/model.py]: The [docs for enum](https://docs.python.org/3/library/enum.html#creating-an-enum) specifically call out that there are no restrictions on the types of values used:

> Member values can be anything: int, str, etc.. If the exact value is unimportant you may use auto instances and an appropriate value will be chosen for you. Care must be taken if you mix auto with other values.

However, it seems that the only place this class is used is in looking up the appropriate class based on the string returned by `status-get`.  There are other approaches we could use to accomplish this, if the enum approach feels off.  For example, we could use [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) on `Status` to register the subclasses automatically.  This would also save us from having to manually maintain the list of available statuses, although we really don't expect this list to ever change anyway.

File: juju/model.py (hunk 5 above)
Comment (johnsca) [reply juju/model.py]: So, I was the one who suggested adding something like `CacheKey` so that we could start distinguishing the parameters that are actually supposed to be part of the key vs ones that just need to get sent on to the constructor but aren't necessarily part of the identity (we already have `cache` as a special case of this).  However, maybe that was a bit of a "swallow the spider to catch the fly" solution, in that the root issue was that this was adding so many new params to the constructor.

`unit_name` is now duplicated, so that's just an artifact of this particular approach.

The "is local unit" flag is only needed to enforce read-only-ness at call time, and if nothing else, we could always just not bother checking that at all in the Python code and let it fail at the Juju level. That feels a bit dirty, but this is a pretty edge case that indicates an error by the charm dev, so maybe it's ok if it lets us keep the code much cleaner.

The backend seems pretty necessary, since we want `model.unit.status` (for the local unit / app at least) to be able to call `status-get` and `status-set`. But, like the cache, it's definitely not part of the identity of the cache entity, so needs to be excluded from the key. However, unlike the `cache` param, it would still need to be passed in from the outside, which makes it more tricky to special-case.

File: juju/model.py (hunk 7 above)
Comment (johnsca) [reply juju/model.py]: Note: To use the `__init_subclass__` approach, you would also need to move the ID (`active`, `blocked`, etc) up to the class level, which actually feels more natural than having it embedded in the constructors.

File: juju/model.py (hunk 6 above)
Comment (johnsca) [reply juju/model.py]: I wonder what the actual trade-off here is in duplication vs complexity with having `EntityStatus` as a separate "generic" descriptor class, vs just using `@property` in the `Application` and `Unit` classes?

File: juju/model.py (hunk 5 above)
Comment (niemeyer) [reply juju/model.py]: The backend indeed feels like a common need, much like the cache itself, which is now being provided into constructors without being part of the cache instance itself. So can just create the model cache with it, and have the cache always handing it as an argument into all the model classes.

For the backend to be able to do the proper actions considering the local unit name (and not trust juju to error), we should move the local unit name origin into the backend itself. After all, it is the one that knows about the local hook world. This will also avoid having to pass the unit name around even into the Model itself as we're doing today.

With those two ideas in place, we should be good to go, and actually simplify our logic a bit with the local unit name behavior. Do I miss some piece of the puzzle?

File: juju/model.py (hunk 6 above)
Comment (dshcherb) [reply juju/model.py]: See #52

File: juju/model.py (hunk 7 above)
Comment (dshcherb) [reply juju/model.py]: Implemented this approach in #52.

**Comment by dshcherb** [comment]
Closing in favor of #52.
